extensions/poetry/main.ts

```diff
@@ -0,0 +1,53 @@
+import { PhylumApi } from "phylum"
+import { parse } from "https://deno.land/std/flags/mod.ts"
+
+// Add packages to the lockfile, and upload it to Phylum for analysis.
+// Return `true` if the packages pass the project's thresholds.
+// Return `false` if the packages don't pass the thresholds, or the processing
+// job is not finished.
+async function poetryCheck(args: string[]) {
+    console.log("Updating package lock…")
+
+    let process = Deno.run({ cmd: ["poetry", "add", '--lock', ...args] })
+    await process.status()
+    await process.close()
+
+    console.log("Package lock updated.\n")
+    console.log("Analyzing packages…")
+
+    const jobId = await PhylumApi.analyze("./poetry.lock")
```

Comment:
So...the entire lockfile is analyzed instead of just the package(s) that were added by the user-supplied `poetry add` or `poetry update` command? That might make it hard to use this extension for any project that does not already meet the risk threshold values.

Reply:
The point of processing the whole lockfile would be addressing the transitive dependencies that arise from `poetry add`, which the user can't know about beforehand, and for which it is not trivial to compute a diff (e.g. new transitive dependencies are easy: just process those; but bumped versions of other transitive dependencies could make it so that the project passes, or stops passing, the thresholds).

If one project does not meet the risk thresholds, that is something the developers need to address, no? Otherwise there is no point in gating `poetry` behind `phylum` in the first place.

If forcing an installation is a desirable use case, we could either add a `--force` flag or tell the user to not run poetry through `phylum`.

This latter option is what I'd gravitate towards, as it "feels wrong" to do and thus may push the user towards doing the right thing instead.

This is not unlike Rust does in some places -- i.e. using `unwrap()` or `unsafe` feels wrong and automatically makes you aware of the fact that you need to be extra careful, which in turn will prompt you to manage errors correctly or put more energy in having the safety invariants hold.

Reply:
Understood. For other readers, this was discussed a bit offline, during an integrations team sync, and the decision was made to continue using the results from analyzing the entire lockfile when making the pass/fail decision.

FWIW, an alternative solution would be to do what the `phylum-ci` tool does, and [get the new dependencies](https://github.com/phylum-dev/phylum-ci/blob/main/src/phylum/ci/ci_base.py#L256-L272) by parsing the lockfile both before and after changes are made to it and then taking the set difference to get the list of new packages. Then, each of those packages' analyze results could be compared to the individual risk domain thresholds set for the project to determine if any single package violates the thresholds.